### ISSUE TITLE
[10.0] Proposal: unify function signatures with the same name

### ISSUE THREAD

**Issue by pqr** [orchidsoftware/platform#1559]
**Is your feature request related to a problem? Please describe.**
Sometimes I'am frustrated with inconsistency in function names and their signatures.

**First example:**
`Field` and `TD` classes have two methods:
1. setter `canSee(bool $value): self` 
2. getter `isSee(): bool`

`Layout` class have only one method: a callback-style getter
`canSee(Repository $query): bool`

I always mix them and forget which is one.

**Second example**
`TD` class: `render(Closure $closure): self`
`Field` class: `render()`

Why I picked `TD` vs `Field`: usually I want to display the same piece of information (model field) on Table/List screen and on Model Details/View screen. I like tweaking render in `TD` - it's easy and straitforward using render callback. But it's not so easy when rendering fields on Details/View screen.

**Additional context**
Changing functions signatures or names is a big breaking change. Not sure if it's worth.

**Comment by tabuna** [comment]
Hi I started to change the visibility functions. Now `Fields/Layouts/TD` and have a common trait.


Without any restrictions. Now you can do this:

```php
Input::make()->canSee(false);
TD::make()->canSee(false);

Layout::rows([])->canSee(false);
```

But now the definition inside the layer is different

```php
// before
public function canSee(Repository $query): bool
{
    return ...;
}

// after
public function isSee(): bool
{
    return ...;
}
```

As for the `render` method, I'm not sure how best to change this.

**Comment by pqr** [comment]
How I see render functions:
1) we need public function for end user to defined a render callback (like current `TD::render(Closure $closure)`)
2) and internal function which actually returns final view or html (like current `Fieldable::render()` and `TD::buildTh()` and `TD::buildTd($repository)`

I like current `TD` naming and propose following naming scheme:
1)  `public function render(Closure $closure)` for end user to define a callback if needed (in both classes: `TD` and `Fieldable`).
2) keep `TD::buildTh()` and `TD::buildTd($repository)`, rename ~~`Fieldable::render()`~~ to `Fieldable::finalRender()` or `Fieldable::buildHtml()`

**Comment by tabuna** [comment]
What should the `render` method do for the field? Transfer function in HTML circuit or only the value?

Perhaps as a modifier rather than a final value?
```php
Input::make('price')->render(function($price) {
  return $price * 72;
})

// output
<input name="price" value="144">
```

Or should we pass everything from query? Or html?

**Comment by pqr** [comment]
One interesting thing about render process:
`TD::buildTd($repository)` accepts `$repository`, so `TD::render(Closure $closure)` - the `$clouse` also accepts `$repository`, means in TD render process we have access to the full model of the currently rendered row.

`Field` class keeps only **value**, not the whole Model, not `$repositry`. 
I propose to pass the $closure in `Fieldable::render(Closure $closure)`.

I found several situations where I need the whole model information during render process, examples:
- I want to render `amount` with red color like this: `if ($model->deviation > $model->amount * 0.1) { /* wrap render result in <span class="danger">...</span> */ }`
- I want to render `quantity` with unit of measure like this `return $model->quantity . ' ' . $model->uom`;

Currently it's easy in table layout having `$repositry` argument in `TD` render `$closure`.

 I think it would be helpful if both `TD` and `Fieldable` render closure will accept `$repositry`.

Example:
```
// For list/table screen
TD::make('quantity')->render(fn($repository) => $model->quantity . ' ' . $model->uom);

// For view screen
Label::make('quantity')->render(fn($repository) => $model->quantity . ' ' . $model->uom);
```

**Comment by pqr** [comment]
> Or should we pass everything from a query? Or HTML?

From my perspective `render(Closure $cloure)` should return HTML (as it currently works in `TD`):

Example of flexibility:
```php
Input::make('quantity')->render(function($repository) {
  $value = $repository->quantity;
  $uom = $resitory->uom; // unit of measure
  return '<input type="text" value="'.$value.'" name="quantity"> <span class="uom">'.$uom.'</span>' . ;
})
```

What I don't like in my example above: I have been forced to create `input` tag manually. Need to think about it.

**Comment by tabuna** [comment]
I took some time to think and came to the following conclusion.
In the column class, we define parameters that do not affect the output, values in a cell for example:

```php
TD::make('updated_at', __('Last edit'))
    ->filter(...)
    ->align(...) 
    ->sort()
    ->render(function (User $user) {
        return $user->updated_at->toDateTimeString();
    }),
```

That is, in this example, the result of calls to the `align`, `sort`, etc. are not replaced by the same `render.` In your example, calling all the previous methods does not make sense, for example:

```php
Input::make('quantity')->required()->render(function($repository) {
  $value = $repository->quantity;
  $uom = $resitory->uom; // unit of measure
  return '<input type="text" value="'.$value.'" name="quantity"> <span class="uom">'.$uom.'</span>' . ;
})
```

There is no point in calling the `required` method. It doesn't even matter which class we use, `Input` or `Checkbox`.

Therefore, I still tend to the closure function returning only the `value` attribute:

```php
Input::make('price')->disabled()->render(function($repository) {
  return $repository->get('price') * $repository->get('course');
});
```

This fits with how TD works and will have the same effects. What do you think about it?

**Comment by pqr** [comment]
> There is no point in calling the required method
Good point!

On the other hand when Field's `render` only calculates the value (like price * course) - this is pointless to have the `render` method. Calculation should be in `query` or in a Model class function `getFinalAmountAttribute` method.

Semantically I feel that `render` method should return the final html representation. At this point I still have no future ideas about useful and valuable `Field::render(Closure $clouse)` implementation.

**Comment by tabuna** [comment]
Since we haven't reached any consensus on the `render` method, version 10.0 was released without any changes.

`canSee`/`isSee` are included in the [released 10.0.0](https://github.com/orchidsoftware/platform/releases/tag/10.0.0)
